skbio/metadata/_mixin.py

```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
         """
         return (self._positional_metadata is not None and
                 len(self.positional_metadata.columns) > 0)
+
+
+class IntervalMetadataMixin(metaclass=abc.ABCMeta):
+    def _interval_metadata_axis_len_(self):
+        ''''''
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def __init__(self, interval_metadata=None):
+        raise NotImplementedError
+
+    def _init_(self, interval_metadata=None):
+        if interval_metadata is None:
+            # Could use deleter but this is less overhead and needs to be fast.
+            self._interval_metadata = None
+        else:
+            # Use setter for validation and copy.
+            self.interval_metadata = interval_metadata
+
+    @property
+    @experimental(as_of="0.5.0-dev")
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
+        '''
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata(
+                self._interval_metadata_axis_len_())
+        return self._interval_metadata
+
+    @interval_metadata.setter
+    def interval_metadata(self, interval_metadata):
+        if isinstance(interval_metadata, IntervalMetadata):
+            upper_bound = interval_metadata.upper_bound
+            axis_len = self._interval_metadata_axis_len_()
+            if upper_bound != axis_len:
```

Comment:
I understand the source of my confusion. `IntervalMetadata` doesn't explicitly define the upper bounds of `locations` or `upper_bound` itself to be inclusive or exclusive, and from the API/usage of the object it isn't clear either. I understood upper bounds to be inclusive for two reasons:

1) The constructor parameter `upper_bound` is documented as:

```
upper_bound : int
        Defines the upper bound of the interval features. No coordinate can
        be greater than it.
```

"No coordinate can be greater than it" implies (to me at least) that a coordinate equal to the upper bound is acceptable.

2) An interval's range(s) are specified as `locations`. `locations` implies coordinates/positions in the coordinate space, so it wasn't clear that the upper bound of a location follows Python's interpretation of slices and ranges.

I understand now that upper bounds are *always* exclusive, and lower bounds are always inclusive, in order to match Python's notion of slices/ranges. There isn't an off-by-one error given this interpretation of bounds.

I can see this potentially causing confusion with users (for example, I've played with the object and reviewed the code, and had an entirely different interpretation of the upper bound). @ebolyen and I talked about this and think it is important at this stage to make the following changes to `IntervalMetadata`/`Interval` in a separate PR in order to make this clearer:

1) Change all occurrences of `locations` to `bounds`. `locations` implies specific positions in the coordinate space, while the name `bounds` is more akin to Python's notion of slices and ranges, which has lower and (exclusive) upper bounds.

2) Change all occurrences of `boundaries` to `fuzzy`. `boundaries` could be misinterpreted to mean inclusive/exclusive for the bounds in `locations`, but it means something entirely different (unsure where the start and/or end really is). "fuzzy" matches Biopython's nomenclature too for their equivalent data structure. "fuzzy" is also a distinctive name that won't get confused with defining inclusive/exclusive bounds, and most users will see the name and ignore it (which covers most use-cases). When `fuzzy` is used, the booleans should be inverted, such that `True` means a lower/upper bound is "fuzzy", and `False` means we have a non-fuzzy (precise) understanding of the lower/upper bound.

3) With inclusive lower bound and exclusive upper bound, an interval with the same start and end (e.g. `[1, 1)`) doesn't represent a position in the coordinate space. These types of intervals should be disallowed to avoid potential confusion (and it isn't clear what exactly it means in the context of interval metadata). If a user wants an interval spanning a single position, use `[1, 2)`.

4) Document `bounds` to always be inclusive lower and exclusive upper. Document `upper_bound` in a similar way.

I think these changes will help make the object more easily understood by users. A couple of my comments on this PR don't apply given the interpretation of upper bounds, I'll follow up on those now.